File: python/55_Jump_Game.py
# coding=utf-8
import unittest
class Solution:
    def solve(self, nums):
        return self.canJump(nums)
        
    def canJump(self, nums):
        if not nums:
            return True
        length=len(nums)
        cur=0
        forward=nums[0]
        while forward<length-1 and forward+nums[forward]<length-1:
            nextpos=forward+nums[forward]
            for i in range(cur,forward):
                if i+nums[i]>nextpos:
                    nextpos=i+nums[i]
            if forward==nextpos:
                return False
            cur=forward
            forward=nextpos
        return True
        
class Test(unittest.TestCase):
    #override
    def setUp(self):
        self.s=Solution()
        
    def tearDown(self):
        print "TestComplete"
        
    def genTestCase(self):
        case=[[2,3,0,0],[1,2,2,0,0],[3,2,1,0,5],[0],[1],[1,0],[1,5,0,1,0]]
        self.cases=case
        
    def test(self):
        self.genTestCase()
        for case in self.cases:
            print "testcase: ", case
            print "result: ",self.s.solve(case)
            # if assert is needed
            
unittest.main()

File: Java/55_Jump_Game.java
public class Solution {
    // 其实这个解的基本思路还是two pointer, 在当前范围去寻找最远可到达的位置, 并不断前移当前范围
    public boolean canJump(int[] nums) {
        if(nums==null||nums.length<=1)return true;
        int cur = 0, next = nums[cur]+cur, len = nums.length;
        int run, cand;
        while(next<len-1&&next!=cur){
            run = cur+1;
            cand = next + nums[next];
            while(run<next){
                cand = Math.max(cand, run+nums[run]);
                run++;
            }
            cur = next;
            next = cand;
        }
        return next>=len-1?true:false;
    }
}